server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -194,30 +196,29 @@ private void verifyResponseReadFully(Header header, TransportResponseHandler<?>
         }
     }
 
-    private <T extends TransportRequest> void handleRequest(TcpChannel channel, Header header, InboundMessage message) throws IOException {
+    private <T extends TransportRequest> void handleRequest(TcpChannel channel, MessageHeader header, InboundMessage message) throws IOException {
         final String action = header.getActionName();
         final long requestId = header.getRequestId();
-        final TransportVersion version = header.getVersion();
         if (header.isHandshake()) {
+            assert header instanceof HandshakeHeader;
             messageListener.onRequestReceived(requestId, action);
             // Cannot short circuit handshakes
             assert message.isShortCircuit() == false;
             final StreamInput stream = namedWriteableStream(message.openOrGetStreamInput());
-            assertRemoteVersion(stream, header.getVersion());
             final TransportChannel transportChannel = new TcpTransportChannel(
                 outboundHandler,
                 channel,
                 action,
                 requestId,
-                version,
+                TransportVersion.CURRENT,   // no message version available, use CURRENT (handshake format hasn't changed in a long time)
                 header.getCompressionScheme(),
                 header.isHandshake(),
                 message.takeBreakerReleaseControl()
             );
             try {
                 handshaker.handleHandshake(transportChannel, requestId, stream);
             } catch (Exception e) {
-                if (TransportVersion.CURRENT.isCompatible(header.getVersion())) {
+                if (((HandshakeHeader)header).getVersion() >= HandshakeHeader.CAN_SEND_ERROR_RESPONSE) {
```

Comment:
I would be happy if we just dropped this. These days we assert that we don't get any deserialization exceptions so the only way that could happen would be if something other than ES were trying to connect, for which it makes no sense to send a response back. I don't think the `channel.sendResponse()` line can throw anything that makes sense to send back either. Or we have a massive bug. In all cases, we should just close the connection.

Reply:
Removed - now it just logs a message